### PR TITLE
Support IndexComponent_slice to ComponentData

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This supports the creation of `IndexedComponent_slice` objects that point to a single object (without any slice or ellipsis).  This supports, e.g. creating a slice to a specific `ComponentData` object.

Using this, the PR simplifies the construction of references to single `ComponentData` objects (this now leverages `IndexedComponent_slice` like other References instead of creating a hidden "floating" Component).

## Changes proposed in this PR:
- Support creating `IndexedComponent_slice` with no Slice / Ellipsis at the initial component
- Update `Reference(ComponentData)` to use `IndexedComponent_slice` objects directly (and not through a hidden floating Component)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
Does this fix #859?

**Comment by jsiirola** [comment]
> Does this fix #859?

No.  #859 is unrelated, and unfortunately expected behavior (all References are Indexed things, regardless of the referent).

File: pyomo/core/base/indexed_component_slice.py [hunk 1]
```diff
@@ -263,11 +264,13 @@ def __str__(self):
...
-                if level[1][3] is not None:
-                    tmp[level[1][3]] = Ellipsis
-                ans += self._getitem_args_to_str([tmp[i] for i in sorted(tmp)])
+                if level[1][1] is not None:
+                    tmp = dict(level[1][1])
+                    tmp.update(level[1][2])
+                    if level[1][3] is not None:
+                        tmp[level[1][3]] = Ellipsis
+                    ans += self._getitem_args_to_str(
+                        [tmp[i] for i in sorted(tmp)])
```
Comment (emma58) [inline pyomo/core/base/indexed_component_slice.py:273]: Do you still need `sorted` here?

**Review by blnicho — APPROVED** [review]

**Review by emma58 — APPROVED** [review]
One question about if you still need `sorted`, but this looks good.

File: pyomo/core/base/indexed_component_slice.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/core/base/indexed_component_slice.py:273]: Yes - I think so: that is putting things back in argument order (the tmp dict maps index position to the fixed / wildcard / ellipsis value)

**Pull request merged**